lib/spack/spack/spec.py

```diff
@@ -1829,6 +1805,43 @@ def concretize(self):
             if s.namespace is None:
                 s.namespace = spack.repo.repo_for_pkg(s.name).namespace
 
+            # Add any patches from the package to the spec.
+            patches = []
+            for cond, patch_list in s.package_class.patches.items():
```

Comment:
@tgamblin @scheibelp There's a thing I don't fully understand about the patch mechanism. The information a packager provides with a directive is stored in a dictionary. This means that the for loop above is not ordered. So we are assuming a weak ordering by constraints. To be more clear, if we have:
```python
patch('foo', when='@1')
patch('another_foo', when='%gcc')
patch('bar', when='@1')
patch('another_bar', when='%gcc')
patch('baz', when='+shared')
```
right now we'll be ensured that:
1. `foo` is applied immediately before `bar`
2. `another_foo` is applied immediately before `another_bar`

because they share the same `cond`, but we don't know whether the first patch is `foo`, `another_foo` or `baz`. 

Would it make sense to relax this constraint and say that the patches must be idempotent? If this is the case, then I think #5596 will solve #5587 after aggregating the two patches for `mpfr` which depend on their relative order of application.

If that's not the case then I think we should go completely the other direction and ensure that the order of patches:
1. reflects the order of directives, regardless of the condition
2. enters the hash mechanism

Note that this is getting a lot trickier than the case before, since we can now inject patches. For example, if in a DAG we have 2 nodes A and B that both want to inject patches on C we must ensure that A and B are always traversed in the same relative order in any graph were they are present.

Do the considerations above sound reasonable to you?

Reply:
To update this, https://github.com/LLNL/spack/pull/5596 now preserves the weak ordering that currently exists on packages.